### ISSUE TITLE
Improve cursor in text editor

### ISSUE THREAD

**Issue by inhalt120g** [godotengine/godot-proposals#8473]
### Describe the project you are working on

A classic arcade game.

### Describe the problem or limitation you are having in your project

The cursor is just too small, it's basically one pixel wide. It's really hard to find on big monitors especially when zoomed out. I know there's also the "caret" options but that one is a bit weak too (again, one pixel width???).

I even set it to blink a bit faster and finding cursor is still a pixel-hunting expedition half of the time.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The cursor should have a few options for thickness.

Additionally, for extra visibility, an option to combine a full caret (as opposed to an outline, like it is now) + single line cursor would be most welcome!!!

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Two pixels thickness:
![image](https://github.com/godotengine/godot-proposals/assets/106015575/9612c7df-2726-427a-a878-af9370c6a49a)
Three pixels thickness:
![image](https://github.com/godotengine/godot-proposals/assets/106015575/9ed7361e-15a0-4813-af8d-9db12d59449f)

Two pixels + full caret! No more chasing one - pixel wide cursor! (the carret uses the same color as cursor at 25 % opaqueness) 
![image](https://github.com/godotengine/godot-proposals/assets/106015575/f33519da-20c2-4bd6-b19a-c24572afe747)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

???

### Is there a reason why this should be core and not an add-on in the asset library?

I'm not sure this can be an add on?

**Comment by Calinou** [comment]
https://github.com/godotengine/godot/pull/58700 addresses the caret width part of this proposal.

As for the second part of the proposal, I haven't seen this in any other editor. If we provide such a setting, I don't know how often it would be used.

**Comment by inhalt120g** [comment]
Yes, 58700 looks like it, I was searching for "cursor" and similar but haven't seen it.

About the second part and how often would it be used:
To my knowledge, sadly there aren't such options elsewhere. I sure wish there were, as a person whose eyesight is not among the best.

Whoever has a setup where letters (and therefore caret) are big, or has super sharp eyes, or knows by heart where his cursor is without even looking at the screen, or accepted and learned to work with "one pixel caret", will probably not find this useful ("this makes no difference, I was able to use the cursor even before, the proposal is worthless"). I expect it will be used as often as there are people whose eyesight is a bit on the weak side or who are more skilled at things other than touch typing or who just like things to be nice and visible.

**Comment by Mickeon** [comment]
Showing the full caret at all times sounds potentially confusing. I get the reasoning, but since the text editor used by Godot itself derives from the TextEdit class, this functionality may be exposed as a property, too. How many would make use of it?

**Comment by inhalt120g** [comment]
"Showing the full caret at all times sounds potentially confusing"
I agree, that's why I didn't propose showing full caret at all times (the proposal is about "line" type cursor, with a potential "subtle highlight" for extra visibility, but the caret itself is still one line). For the same reason I also didn't propose the chunky "DOS terminal" type cursor, which has a nice visibility but it's slightly illogical that it's adding characters behind itself and not where it currently is.

**Comment by inhalt120g** [comment]
There's a setting that looks like it's meant to do what the second part of the proposal is suggesting:
![image](https://github.com/godotengine/godot-proposals/assets/106015575/fbabb1c2-b730-46c6-8392-2e3d9b636fda)
…but it's currently unused so I can't check it:
![image](https://github.com/godotengine/godot-proposals/assets/106015575/c22d414a-af72-4ed7-8cf9-6ede02e864d8)

**Comment by Calinou** [comment]
@Paulb23 @bruvzg Do you know what `caret_background_color` was supposed to look like in the script editor?

**Comment by bruvzg** [comment]
> Do you know what caret_background_color was supposed to look like in the script editor?

No idea, it seems to be unused.

> Two pixels + full caret! No more chasing one - pixel wide cursor! (the carret uses the same color as cursor at 25 % opaqueness)

There's already support for block caret, so adding support for something like this should not be hard. But note that caret in the bidirectional text can have additional forms (split caret + direction markers, and ligature caret) which should be accounted for as well.

<img width="341" alt="caret_forms" src="https://user-images.githubusercontent.com/7645683/194474871-ca63c8eb-37c5-47b0-9e27-a7f9f9ec9bc4.png">

**Comment by inhalt120g** [comment]
The way I imagined the "highlight" (I wrongly called it "full caret" in the original proposal) to behave is exactly the same like what the "block" type cursor does in the current version (so I assume all the logic for all languages is already implemented). 

Including, for example, how the block-type cursor actually covers the tab when the cursor is on the left from the place where the tab is (although in the current implementation it's a bit confusing because it's not clear at a glance where the actual cursor is). The "highlight" part of the cursor could inherit all the behaviors from the current block-type, only drawn solid as opposed to current 1-pixel outline-only (same for the "main" part of the cursor, it should continue behaving the same when drawn a bit thicker).

I don't know how the block-type cursor behaves in bidirectional text but in all the contexts I tried it with (GDScript, Chinese, Japanese) it worked well.

**Comment by Paulb23** [comment]
`caret_background_color` should be the colour of the text behind the caret when block caret is enabled, but currently there's a bug where it's not actually used anywhere.

**Comment by inhalt120g** [comment]
If this is a bug, should I file it under issues?

**Comment by Calinou** [comment]
> If this is a bug, should I file it under issues?

Yes, I suggest opening an issue for what Paulb23 pointed out :slightly_smiling_face:

**Comment by inhalt120g** [comment]
Done: https://github.com/godotengine/godot/issues/86281
